### PR TITLE
Optimize distinct_permutations

### PR DESCRIPTION
**Pull request by bbayles**
@jferard, this PR implements a variant of the change you proposed in PR #208. Any thoughts? Looks like the SO poster didn't respond.

I moved the helper function body into the loop. I think it makes it a little clearer that we're building up the results in memory.

### CODE REVIEW

**Comment by jferard** [comment]
Hi @bbayles! I had forgotten this old PR. As you pointed out, there is still no answer from the SO poster (although he has some activity).

The version you wrote is closer to the SO original post than mine, because I chose not to build the permutation *list* for every `i < len(iterable)`. My idea was to keep it lazy: distinct_permutations of `[e1, e2, ..., en]` needs distinct permutations of `[e1, e2, ..., en-1]` and `en`, but doesn't need the list of those permutation to be built and stored. We can create the elements on the fly.

I made some benchmarks to compare both versions. I give you the full code:

# `distinct_time.py`

```
import string
import random
import itertools
import timeit

def distinct_permutations(iterable):
    all_perms = [[]]
    for item in iterable:
        new_perms = []
        for perm in all_perms:
            for i in range(len(perm)):
                new_perms.append(perm[:i] + [item] + perm[i:])
                if perm[i] == item:
                    break
            else:  # No break
                new_perms.append(perm + [item])

        all_perms = new_perms

    return (tuple(perm) for perm in all_perms)

def distinct_permutations_jferard(iterable):
    def distinct_permutations_generator(permutations_generator, e):
        for permutation in permutations_generator:
            for j in range(len(permutation)):
                yield permutation[:j] + [e] + permutation[j:]
                if permutation[j] == e:
                    break
            else:
                yield permutation + [e]


    permutations_generator = [[]]
    for e in iterable:
        permutations_generator = distinct_permutations_generator(
            permutations_generator, e)

    return (tuple(t) for t in permutations_generator)

def test_func(func, values):
    def test_func_closure():
        return list(func(values))
    return test_func_closure

times = {}
for _ in range(100):
    values = [random.choice(string.ascii_letters[:5]) for _ in range(10)]
    result = set(itertools.permutations(values))
    for func in [distinct_permutations_jferard, distinct_permutations]:
        t = test_func(func, values)
        assert sorted(t()) == sorted(result), "expected: {}, actual: {} ({})".format(result, t(), func)
        times.setdefault(func, []).append(timeit.timeit(t, number=20))

for k,v in times.items():
    print (k, sum(v))
```

## Results:

```
jferard@jferard-Z170XP-SLI:~/prog/python$ python3 distinct_time.py 
<function distinct_permutations_jferard at 0x7f083c1056a8> 38.9083186589869
<function distinct_permutations at 0x7f083c1128c8> 42.55286032900767
```

As you can see, nothing spectacular, but my version is slightly faster (I made this tests more than once). Obviously, if you replace the line 

    return list(func(values))

in the closure by:

    return list(itertools.islice(func(values), 10))

The results are different (but don't ask me any use case for this):

```
jferard@jferard-Z170XP-SLI:~/prog/python$ python3 distinct_time.py 
<function distinct_permutations_jferard at 0x7fa1301d56a8> 0.026467208999747527
<function distinct_permutations at 0x7fa1301e28c8> 36.471522780007945
```

# `distinct_memory.py`

```
import string
import random
import itertools
import sys

@profile
def distinct_permutations(iterable):
    all_perms = [[]]
    for item in iterable:
        new_perms = []
        for perm in all_perms:
            for i in range(len(perm)):
                new_perms.append(perm[:i] + [item] + perm[i:])
                if perm[i] == item:
                    break
            else:  # No break
                new_perms.append(perm + [item])

        all_perms = new_perms

    return (tuple(perm) for perm in all_perms)

@profile
def distinct_permutations_jferard(iterable):
    def distinct_permutations_generator(permutations_generator, e):
        for permutation in permutations_generator:
            for j in range(len(permutation)):
                yield permutation[:j] + [e] + permutation[j:]
                if permutation[j] == e:
                    break
            else:
                yield permutation + [e]


    permutations_generator = [[]]
    for e in iterable:
        permutations_generator = distinct_permutations_generator(
            permutations_generator, e)

    return (tuple(t) for t in permutations_generator)

def test_func(func, values):
    def test_func_closure():
        return list(func(values))
    return test_func_closure

values = [random.choice(string.ascii_letters[:5]) for _ in range(int(sys.argv[1]))]
#    result = set(itertools.permutations(values))
for func in [distinct_permutations_jferard, distinct_permutations]:
    t = test_func(func, values)
#        assert sorted(t()) == sorted(result), "expected: {}, actual: {} ({})".format(result, t(), func)
    t()
```

## Results

For 10 elements:
```
jferard@jferard-Z170XP-SLI:~/prog/python$ python -m memory_profiler distinct_memory.py 10
Filename: distinct_memory.py

Line #    Mem usage    Increment   Line Contents
================================================
     6   33.168 MiB   33.168 MiB   @profile
     7                             def distinct_permutations(iterable):
     8   33.168 MiB    0.000 MiB       all_perms = [[]]
     9   40.828 MiB    0.000 MiB       for item in iterable:
    10   35.645 MiB    0.000 MiB           new_perms = []
    11   41.461 MiB    0.258 MiB           for perm in all_perms:
    12   41.461 MiB    0.254 MiB               for i in range(len(perm)):
    13   41.461 MiB    0.258 MiB                   new_perms.append(perm[:i] + [item] + perm[i:])
    14   41.461 MiB    0.258 MiB                   if perm[i] == item:
    15   41.461 MiB    0.004 MiB                       break
    16                                         else:  # No break
    17   33.168 MiB    0.000 MiB                   new_perms.append(perm + [item])
    18                             
    19   40.828 MiB    0.000 MiB           all_perms = new_perms
    20                             
    21   40.828 MiB    0.000 MiB       return (tuple(perm) for perm in all_perms)


Filename: distinct_memory.py

Line #    Mem usage    Increment   Line Contents
================================================
    23   32.348 MiB   32.348 MiB   @profile
    24                             def distinct_permutations_jferard(iterable):
    25   32.348 MiB    0.000 MiB       def distinct_permutations_generator(permutations_generator, e):
    26                                     for permutation in permutations_generator:
    27                                         for j in range(len(permutation)):
    28                                             yield permutation[:j] + [e] + permutation[j:]
    29                                             if permutation[j] == e:
    30                                                 break
    31                                         else:
    32                                             yield permutation + [e]
    33                             
    34                             
    35   32.348 MiB    0.000 MiB       permutations_generator = [[]]
    36   32.348 MiB    0.000 MiB       for e in iterable:
    37   32.348 MiB    0.000 MiB           permutations_generator = distinct_permutations_generator(
    38   32.348 MiB    0.000 MiB               permutations_generator, e)
    39                             
    40   32.348 MiB    0.000 MiB       return (tuple(t) for t in permutations_generator)
```

And for 13 elements:

```
jferard@jferard-Z170XP-SLI:~/prog/python$ python -m memory_profiler distinct_memory.py 13
Filename: distinct_memory.py

Line #    Mem usage    Increment   Line Contents
================================================
     6   33.332 MiB   33.332 MiB   @profile
     7                             def distinct_permutations(iterable):
     8   33.332 MiB    0.000 MiB       all_perms = [[]]
     9  207.191 MiB    0.000 MiB       for item in iterable:
    10   98.223 MiB    0.000 MiB           new_perms = []
    11  225.621 MiB    0.258 MiB           for perm in all_perms:
    12  225.621 MiB    0.258 MiB               for i in range(len(perm)):
    13  225.621 MiB    5.156 MiB                   new_perms.append(perm[:i] + [item] + perm[i:])
    14  225.621 MiB    0.258 MiB                   if perm[i] == item:
    15  225.621 MiB    0.258 MiB                       break
    16                                         else:  # No break
    17   33.586 MiB    0.000 MiB                   new_perms.append(perm + [item])
    18                             
    19  207.191 MiB    0.000 MiB           all_perms = new_perms
    20                             
    21  207.191 MiB    0.000 MiB       return (tuple(perm) for perm in all_perms)


Filename: distinct_memory.py

Line #    Mem usage    Increment   Line Contents
================================================
    23   32.492 MiB   32.492 MiB   @profile
    24                             def distinct_permutations_jferard(iterable):
    25   32.492 MiB    0.000 MiB       def distinct_permutations_generator(permutations_generator, e):
    26                                     for permutation in permutations_generator:
    27                                         for j in range(len(permutation)):
    28                                             yield permutation[:j] + [e] + permutation[j:]
    29                                             if permutation[j] == e:
    30                                                 break
    31                                         else:
    32                                             yield permutation + [e]
    33                             
    34                             
    35   32.492 MiB    0.000 MiB       permutations_generator = [[]]
    36   32.492 MiB    0.000 MiB       for e in iterable:
    37   32.492 MiB    0.000 MiB           permutations_generator = distinct_permutations_generator(
    38   32.492 MiB    0.000 MiB               permutations_generator, e)
    39                             
    40   32.492 MiB    0.000 MiB       return (tuple(t) for t in permutations_generator)
```

As you can see, the memory usage is (I guess) constant in my version.

To summarize, I think my version is better, but the awful truth is, since the algorithm is (and will remain whatever you do) *O(2^n)*, I don't now if this is really important!

**Comment by bbayles** [comment]
Ah, thanks for the detailed response. Yes, I can see now how yours has a further optimization. I'll clean up the errors and merge yours.
